crates/sc-proof-of-time/src/pot_state.rs

```diff
@@ -0,0 +1,316 @@
+                self.add_to_tip(&mut state, proof, ExtendAction::MergeFutureProofs);
+                return Ok(());
+            }
+        };
+
+        // Case 1: the proof is for an older slot
+        if proof.slot_number <= tip.slot_number {
+            return Err(PotStateError::StaleProof {
+                tip_slot: tip.slot_number,
+                proof_slot: proof.slot_number,
+            });
+        }
+
+        // Case 2: the proof extends the tip
+        if (tip.slot_number + 1) == proof.slot_number {
+            let expected_seed = tip.next_seed(None);
+            if proof.seed != expected_seed {
+                return Err(PotStateError::InvalidSeed {
+                    expected: expected_seed,
+                    actual: proof.seed,
+                });
+            }
+
+            let expected_key = tip.next_key();
+            if proof.key != expected_key {
+                return Err(PotStateError::InvalidKey {
+                    expected: expected_key,
+                    actual: proof.key,
+                });
+            }
+
+            self.proof_of_time
+                .verify(proof)
+                .map_err(PotStateError::InvalidProof)?;
+
+            // All checks passed, advance the tip with the new proof
+            self.add_to_tip(&mut state, proof, ExtendAction::MergeFutureProofs);
+            return Ok(());
+        }
+
+        // Case 3: proof for a future slot
+        self.handle_future_proof(&mut state, &tip, sender, proof)
+    }
+
+    /// Handles the received proof for a future slot.
+    fn handle_future_proof(
+        &self,
+        state: &mut InternalState,
+        tip: &PotProof,
+        sender: PeerId,
+        proof: &PotProof,
+    ) -> Result<(), PotStateError> {
+        // Reject if too much into future
+        if (proof.slot_number - tip.slot_number) > self.config.max_future_slots {
+            return Err(PotStateError::TooFuturistic {
+                tip_slot: tip.slot_number,
+                proof_slot: proof.slot_number,
+            });
+        }
+
+        match state.future_proofs.entry(proof.slot_number) {
+            Entry::Vacant(entry) => {
+                let mut proofs = BTreeMap::new();
+                proofs.insert(sender, proof.clone());
+                entry.insert(proofs);
+                Ok(())
+            }
+            Entry::Occupied(mut entry) => {
+                let proofs_for_slot = entry.get_mut();
+                // Reject if the sender already sent a proof for same slot number.
+                if proofs_for_slot.contains_key(&sender) {
+                    return Err(PotStateError::DuplicateProofFromPeer(sender));
+                }
+
+                // Reject if the there is an existing proof with same
+                // seed/key but different checkpoints.
+                if let Some(existing_proof) =
+                    proofs_for_slot.values().find(|p| p.seed == proof.seed)
+                {
```

Comment:
It shouldn't do this check at all at this point because we don't know what the seed and key should be for future proof and aren't guaranteed that the first one we received is correct. See my [comment](https://github.com/subspace/subspace/pull/1659#discussion_r1262720441) on previous PR